### ISSUE TITLE
Ability to specify custom solver output text

### ISSUE THREAD

**Issue by danpetry** [conda/conda-libmamba-solver#314]
### Checklist

- [X] I added a descriptive title
- [X] I searched open requests and couldn't find a duplicate

### What is the idea?

I would like to be able to tell the user of a package, in natural language, why their package isn't installable in the case that the solver can't resolve it. In most cases, it's appropriate to describe the version conflicts in the way that the solver already does. However, in some cases, additional information, or information given in a way that doesn't require knowledge of dependency management, would be helpful.

### Why is this needed?

I have cpu and gpu variants of a package, `pytorch`. The GPU variant depends on certain versions of virtual packages (`_CUDA` on linux and `_OSX` on mac). I have a metapackage `pytorch-gpu` which will explicitly request the GPU version and fail when the system doesn’t support it. The solver output has the info in there, it’s true, but at the moment it will say something like “unsolvablity error: pytorch-gpu -> _CUDA == 11.7 or 12.1” Whereas I’d like to say “You have requested the GPU package, but your system doesn’t support it, because you’ve only got CUDA version 11.3. Please upgrade your system CUDA version to 11.7 or 12.1".

Another case is for enterprise users, where the company filter has removed the package from the on-prem repo. It would be nice to be able to tell the company devs this.

### What should happen?

There would be an obvious message, probably near the unsatisfiability error that's currently given.

**Comment by costrouc** [comment]
@danpetry thanks for opening an issue! 

Recently the error messages for unsolvable specs has improved a lot from conda-libmamba-solver using mamba/libmamba's unsolvable messages. There definitely is still a lot of room to improve.

I think this could be difficult to handle since we'd have to build in heuristics about why something didn't work/solve. I agree that the gpu/cpu solutions are a tricky one.

**Comment by danpetry** [comment]
Forgive my ignorance, but what do you mean by heuristics? I thought already there's functionality in there to determine why something didn't solve (as described by the current error messages)?

I was thinking about being able to specify an error message in the recipe? Something like:
```
- requirements:
  - run:
    - _CUDA >=11.8 # %Pytorch-gpu only supports CUDA v 11.8 and above. Please upgrade your CUDA version to install this package.
```
Where `%` denotes the error message that should be displayed if this requirement is causing the error. That could address the first use case above. I guess for the second you'd need some kind of on-prem register of which packages have been blocked and to try to solve with those too. Not sure that would belong in the solver, possibly in some higher-level tool which calls the solver, or a post-command hook conda plugin..?

**Comment by jaimergp** [comment]
Hey Dan, thanks for the proposal. Unfortunately, this feature you are requesting is sadly out of the scope of this particular repository:

* The conflict errors are produced by `libmamba` over at `mamba-org/mamba`. That's C++ code so we can't do much from the Python side other than parse and reinterpret a bit.
* The annotations you are suggesting in the recipe would need to be saved somewhere in the package metadata. This would require new additions to the metadata standards in repodata.json, in a backwards-compatible way (e.g. `depends_annotations`).
* We are trying to get away from semantic comments for the new `meta.yaml` iteration. For example, a more desirable syntax would be:

```
requirements:
  run:
    - annotated:
        spec: _CUDA >=11.8
        annotation: Pytorch-gpu only supports CUDA v 11.8 and above. Please upgrade your CUDA version to install this package.
```

Then that metadata would need to be consumed by the solver backend when producing or delivering the messages.

Anyway, to summarize, this is a big lift that requires ecosystem-wide changes in the distributed metadata and as such it's probably best discussed via a formal proposal in `conda-incubator/ceps`.

I'll close here because we can't do much about it, but I am happy to participate in the CEP conversation if that happens.

**Comment by danpetry** [comment]
Ok, thanks for the explanation!
